op-batcher/batcher/driver.go

```diff
@@ -542,92 +541,137 @@ func (l *BatchSubmitter) receiptsLoop(wg *sync.WaitGroup, receiptsCh chan txmgr.
+			l.Log.Error("SetMaxDASize RPC method unavailable on endpoint, shutting down. Either enable it or disable throttling.",
+				"endpoint", endpoint, "err", err)
+
+			// We have a strict requirement that all endpoints must have the SetMaxDASize endpoint, and shut down if this RPC method is not available
 			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 			defer cancel()
 			// Call StopBatchSubmitting in another goroutine to avoid deadlock.
 			go func() {
-				// Always returns nil. An error is only returned to expose this function as an RPC.
 				_ = l.StopBatchSubmitting(ctx)
 			}()
 			return
 		} else if err != nil {
-			l.Log.Error("SetMaxDASize rpc failed, retrying.", "err", err)
+			l.Log.Error("SetMaxDASize RPC failed for endpoint, retrying.", "endpoint", endpoint, "err", err)
 			retryTimer.Reset(retryInterval)
 			return
 		}
+
 		if !success {
-			l.Log.Error("Result of SetMaxDASize was false, retrying.")
+			l.Log.Error("Result of SetMaxDASize was false for endpoint, retrying.", "endpoint", endpoint)
 			retryTimer.Reset(retryInterval)
+			return
 		}
+
+		l.Log.Info("Successfully set max DA size on endpoint",
+			"endpoint", endpoint,
+			"max_tx_size", maxTxSize,
+			"max_block_size", maxBlockSize)
 	}
 
 	cachedPendingBytes := int64(0)
-
 	for {
 		select {
-		case pendingBytes, ok := <-pendingBytesUpdated:
+		case pendingBytes, ok := <-endpointUpdates:
 			if !ok {
 				// If the channel was closed, this is our signal to exit
-				l.Log.Info("throttlingLoop returning")
+				l.Log.Info("Endpoint throttling loop shutting down", "endpoint", endpoint)
 				return
 			}
 			updateParams(pendingBytes)
 			cachedPendingBytes = pendingBytes
-		case <-l.ActiveSeqChanged:
-			updateParams(cachedPendingBytes)
 		case <-retryTimer.C:
 			updateParams(cachedPendingBytes)
 		}
 	}
 }
 
+// throttlingLoop acts as a distributor that spawns individual throttling loops for each endpoint
+// and fans out the pending bytes updates to each endpoint
+func (l *BatchSubmitter) throttlingLoop(wg *sync.WaitGroup, pendingBytesUpdated chan int64) {
+	defer wg.Done()
+	l.Log.Info("Starting DA throttling loop")
+	updateChans := make([]chan int64, len(l.Config.ThrottlingEndpoints))
+
+	innerWg := sync.WaitGroup{}
+
+	// For each configured endpoint
+	for i, endpoint := range l.Config.ThrottlingEndpoints {
+		// Create channel for this endpoint
+		updateChans[i] = make(chan int64, 1)
+
+		// Start throttling loop for this endpoint
+		innerWg.Add(1)
+		go l.singleEndpointThrottler(&innerWg, updateChans[i], endpoint)
+		l.Log.Info("Started throttling loop for endpoint", "endpoint", endpoint)
+	}
+
+	// fan out events
+	for pb := range pendingBytesUpdated {
+		for i, updateChan := range updateChans {
+			// Send pending bytes to each endpoint's channel
+			select {
+			case updateChan <- pb:
+			default:
+				l.Log.Warn("Throttling loop: channel full, skipping update", "endpoint", l.Config.ThrottlingEndpoints[i])
+			}
```

Comment:
I like that. If one throttling goroutine is busy, e.g. redialing the endpoint, the channel won't be spammed with many updates.

OTOH, ideally each routine always throttles with the _latest value_ once it starts sending throttling updates... but this isn't as easy to accomplish. We'd need to work with a signaling channel plus a shared atomic value or so. So this routine could maintain updating the atomic, while the throttling loops read from it right before sending the throttling update to their endpoint. Wdyt?

Reply:
Not directly related to your point here, but I refactored things such that the distributor routine does the threshold comparison and just sends a bool to each individual throttling routine [6c7c1b5](https://github.com/ethereum-optimism/optimism/pull/15603/commits/6c7c1b521e160bd1111d6f10f948fa4e4c1b45e4).

Reply:
Went ahead with the "latest value" idea here [7b2d77b](https://github.com/ethereum-optimism/optimism/pull/15603/commits/7b2d77b80fbc7d400b70446fc14f9b52a2b3195a)